pkg/metastore/ms_client.go

```diff
@@ -0,0 +1,275 @@
+	GetDatabase(ctx context.Context, name string) (r *Database, err error)
+	GetDatabases(ctx context.Context, pattern string) (databases []*Database, err error)
+	GetTables(ctx context.Context, dbName string, pattern string) (tables []*Table, err error)
+}
+
+type WriteClient interface {
+	CreateTable(ctx context.Context, tbl *Table) (err error)
+	AlterTable(ctx context.Context, dbName string, tableName string, newTable *Table) (err error)
+	AddPartitions(ctx context.Context, tableName string, dbName string, newParts []*Partition) (err error)
+	AlterPartitions(ctx context.Context, dbName string, tableName string, newPartitions []*Partition) (err error)
+	AlterPartition(ctx context.Context, dbName string, tableName string, values *Partition) (err error)
+	AddPartition(ctx context.Context, tableName string, dbName string, newPartition *Partition) (err error)
+	DropPartition(ctx context.Context, dbName string, tableName string, values []string) (err error)
+	CreateDatabaseIfNotExists(ctx context.Context, database *Database) (err error)
+	NormalizeDBName(name string) string // NormalizeDBName changes the db name to be a valid name for the client
+}
+
+type Client interface {
+	ReadClient
+	WriteClient
+}
+
+func CopyOrMerge(ctx context.Context, fromClient Client, toClient Client, fromDB, fromTable, toDB, toTable, toBranch, serde string, partition []string) error {
+	transformLocation := func(location string) (string, error) {
+		return ReplaceBranchName(location, toBranch)
+	}
+	return copyOrMerge(ctx, fromClient, toClient, fromDB, fromTable, toDB, toTable, serde, partition, transformLocation)
+}
+
+func copyOrMerge(ctx context.Context, fromClient, toClient Client, fromDB, fromTable, toDB, toTable, serde string, partition []string, transformLocation func(location string) (string, error)) error {
+	if len(partition) > 0 {
+		return CopyPartition(ctx, fromClient, toClient, fromDB, fromTable, toDB, toTable, serde, partition, transformLocation)
+	}
+	hasTable, err := toClient.HasTable(ctx, toDB, toTable)
+	if err != nil {
+		return err
+	}
+	if !hasTable {
+		table, err := fromClient.GetTable(ctx, fromDB, fromTable)
+		if err != nil {
+			return err
+		}
+		partitions, err := fromClient.GetPartitions(ctx, fromDB, fromTable)
+		if err != nil {
+			return err
+		}
+		return Copy(ctx, table, partitions, toDB, toTable, serde, toClient, transformLocation)
+	}
+	table, err := fromClient.GetTable(ctx, fromDB, fromTable)
+	if err != nil {
+		return err
+	}
+	partitions, err := fromClient.GetPartitions(ctx, fromDB, fromTable)
+	if err != nil {
+		return err
+	}
+	partitionCollection := NewPartitionCollection(partitions)
+	return Merge(ctx, table, partitionCollection, toDB, toTable, serde, toClient, transformLocation)
+}
+
+func CopyOrMergeFromValues(ctx context.Context, fromClient Client, fTable *Table, toClient Client, fromDB, fromTable, toDB, toTable, serde string, transformLocation func(location string) (string, error)) error {
```

Comment:
general note - passing a func for transformation which we do not use for anything else to configure the behavior makes the implementation hard to read.
looks like the operation should hold a state/struct that would have to hold the relevant information for the copy operation it would remove the need for this function.